help_function/src/onedpl_test_tagged_pointer.cpp

```diff
@@ -0,0 +1,158 @@
+
+  dpct::tagged_pointer<SystemTag, int> int_ptr_end = int_ptr_beg + n;
+  failing_tests += ASSERT_EQUAL(
+      int_ptr_name + " add operator",
+      static_cast<int *>(int_ptr_end) - static_cast<int *>(int_ptr_beg), n);
+
+  dpct::tagged_pointer<SystemTag, int> expect_beg = int_ptr_end - n;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subtract operator",
+                                int_ptr_beg == expect_beg, true);
+
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " difference operator",
+                                int_ptr_end - int_ptr_beg, n);
+
+  expect_beg++;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " postfix increment",
+                   (int_ptr_beg + 1) == expect_beg, true);
+
+  expect_beg--;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " postfix decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  ++expect_beg;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " prefix increment",
+                   (int_ptr_beg + 1) == expect_beg, true);
+  --expect_beg;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " prefix decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  expect_beg += 2;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " addition assignment",
+                   (int_ptr_beg + 2) == expect_beg, true);
+
+  expect_beg -= 2;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " subtraction assignment",
+                   int_ptr_beg == expect_beg, true);
+
+  // Test conversion to base pointer
+  int *int_ptr_beg_raw = int_ptr_beg;
+  int *int_ptr_end_raw = int_ptr_end;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " conversion to int*",
+                                int_ptr_end_raw - int_ptr_beg_raw, n);
+
+  // device allocations use malloc_shared so this is safe
+  *int_ptr_beg = 4;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " dereference operator",
+                                *int_ptr_beg == 4, true);
+  int_ptr_beg[1] = 2;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subscript operator",
+                                int_ptr_beg[1] == 2, true);
+
+  dpct::tagged_pointer<SystemTag, integer_wrapper> int_wrapper_beg =
+      dpct::malloc<integer_wrapper>(system, 1);
+  int_wrapper_beg->val = 5;
+  failing_tests += ASSERT_EQUAL(int_wrapper_name + " arrow operator",
+                                (*int_wrapper_beg).val == 5, true);
+
+  dpct::free(system, void_ptr_beg);
+  dpct::free(system, int_wrapper_beg);
+  return failing_tests;
+}
+
+template <typename SystemTag> int test_tagged_pointer_iteration(void) {
+  constexpr ::std::size_t n = 1024;
+  int return_fail_code = 0;
+  SystemTag system;
+  bool is_host = ::std::is_same_v<dpct::host_sys_tag, SystemTag>;
+  ::std::string sys = is_host ? "dpct::host_sys_tag" : "dpct::device_sys_tag";
+  std::string int_ptr_name = "dpct::tagged_pointer<" + sys + ", int>";
+
+  auto policy = oneapi::dpl::execution::dpcpp_default;
```

Comment:
I believe this crashes for `host_sys_tag` when run on a GPU.  The CI seems to have caught it with a RUNFAIL.  
This makes sense if we are running a device policy fill on host located data.  

It may be good to pass and use a policy, and let that determine the location of the memory.

Reply:
Good catch, I will make this change.

Reply:
I have made this change (and some stylistic clang-format changes). This fixes the bug locally.

CI is rerunning to confirm the issue is fixed.